File: BosaiApp/BosaiApp/Models/Shelter.swift
import Foundation
import GRDB

struct Shelter: Codable, FetchableRecord, PersistableRecord, Identifiable {
    static let databaseTableName = "shelters"

    var id: String { shelterId }
    var shelterId: String
    var name: String
    var lat: Double
    var lon: Double
    var elevationM: Double
    var coastDistanceM: Double
    var types: String          // JSON配列: ["earthquake","tsunami","fire"]
    var capacity: Int?
    var isOpenSpace: Bool

    enum Columns: String, ColumnExpression {
        case shelterId = "shelter_id"
        case name, lat, lon
        case elevationM = "elevation_m"
        case coastDistanceM = "coast_distance_m"
        case types, capacity
        case isOpenSpace = "is_open_space"
    }

    enum CodingKeys: String, CodingKey {
        case shelterId = "shelter_id"
        case name, lat, lon
        case elevationM = "elevation_m"
        case coastDistanceM = "coast_distance_m"
        case types, capacity
        case isOpenSpace = "is_open_space"
    }

    // MARK: - Helper

    var typeArray: [String] {
        guard let data = types.data(using: .utf8),
              let arr = try? JSONDecoder().decode([String].self, from: data) else { return [] }
        return arr
    }
}

File: bosai_app/lib/models/shelter.dart
/// 避難所モデル（設計書 §4.2 のスキーマに対応）
class Shelter {
  final String shelterId;
  final String name;
  final double lat;
  final double lon;
  final double elevationM; // 海抜
  final double coastDistanceM; // 海岸からの距離
  final String types; // カンマ区切り: "earthquake,tsunami,fire"
  final int capacity;

  const Shelter({
    required this.shelterId,
    required this.name,
    required this.lat,
    required this.lon,
    required this.elevationM,
    required this.coastDistanceM,
    required this.types,
    required this.capacity,
  });

  bool supports(String type) => types.split(',').contains(type);

  factory Shelter.fromMap(Map<String, dynamic> map) => Shelter(
        shelterId: map['shelter_id'] as String,
        name: map['name'] as String,
        lat: (map['lat'] as num).toDouble(),
        lon: (map['lon'] as num).toDouble(),
        elevationM: (map['elevation_m'] as num).toDouble(),
        coastDistanceM: (map['coast_distance_m'] as num).toDouble(),
        types: map['types'] as String,
        capacity: map['capacity'] as int,
      );

  Map<String, dynamic> toMap() => {
        'shelter_id': shelterId,
        'name': name,
        'lat': lat,
        'lon': lon,
        'elevation_m': elevationM,
        'coast_distance_m': coastDistanceM,
        'types': types,
        'capacity': capacity,
      };
}
